### PR TITLE
Third party mocha reporter warning suppression

### PR DESCRIPTION
**Pull request by scottlabs**
Try loading third party mocha reporter first, and on failure fall back to passing as a string

### CODE REVIEW

File: bin/cli.js [hunk 1]
```diff
@@ -90,35 +90,40 @@ mocha.setup({
...
+if (typeof process === 'undefined') {
+  // a poor node.js process shim - totally not great
+  var sys = require('system')
+  this.process = {
+    pid: sys.pid,
+    env: sys.env,
+    argv: sys.args.splice(),
+    stdin: sys.stdin,
+    stdout: sys.stdout,
+    stderr: sys.stderr
```
Comment (nathanboktae) [inline bin/cli.js]: This process shim needs to be created before trying to require a 3rd party reporter. Hmm a test should have failed for you here.... Coverage is missing.

File: bin/cli.js [hunk 2]
```diff
@@ -90,35 +90,40 @@ mocha.setup({
...
+  } catch (e) {
+    reporter = opts.reporter
   }
+}
 
-  // I don't want to use isAbsolute here as it could be a node module or a relative path
-  if (opts.reporter.indexOf('.') === 0) {
-    opts.reporter = fs.absolute(opts.reporter)
+// phantomjs exits immediately if it can't find a module due to exitOnError: true from above... that should probably be false
+// either way a module may lazily require something and fail later, so a try/catch with an informative message isn't possible now.
```
Comment (nathanboktae) [inline bin/cli.js]: This PR basically disproves this comment :) hmm maybe a behavior change in [casper's `patchRequire`](https://github.com/n1k0/casperjs/blob/a2c47c0452b5a1ede70971a098cc665b49c3b2ca/bin/bootstrap.js#L200-L210). Please delete this comment then.

**Comment by nathanboktae** [comment]
I like the PR - if you could fix those 2 issues that'd be great. Thanks Scott for your contributions :)

**Comment by scottlabs** [comment]
One more comment - I added a try / catch block around the reporter require. Not sure if this is the best way to do it. 

Basically, errors in third party reporters were causing cli.js to hang indefinitely, so its a way around it. Kinda sucks having two try / catches one after another though.

**Comment by nathanboktae** [comment]
> Kinda sucks having two try / catches one after another though.

Yeah true. But we should exit though rather than swallow it, I think.

File: test/all.coffee [hunk 3]
```diff
@@ -365,6 +368,17 @@ describe 'mocha-casperjs', ->
...
+      runMochaCasperJsTest
+        params: ['--reporter=./test/process-reporter']
+        test: ->
+          1.should.be.ok
+      , (output, code) ->
+        output.indexOf("Can\'t find variable: process").should.equal(-1)
+        output.indexOf('stdout').should.be.above(-1)
+        output.indexOf('stderr').should.be.above(-1)
+        done()
+
```
Comment (nathanboktae) [inline test/all.coffee]: :+1: thanks!

**Comment by scottlabs** [comment]
Good call on exiting instead of swallowing. Do you follow guidelines in terms of passing an exit code to casper.exit?

**Comment by nathanboktae** [comment]
> Do you follow guidelines in terms of passing an exit code to casper.exit?

Wasn't aware of any. Mocha's exit code is equal to the number of tests failed, so going negative is best.

**Pull request merged**